alchemiscale_fah/protocols/protocolunit.py

```diff
@@ -0,0 +1,284 @@
+"""
+:mod:`alchemiscale_fah.protocols.protocolunit` --- reusable ProtocolUnits for Folding@Home protocols
+====================================================================================================
+
+"""
+
+import abc
+from typing import List, Tuple, Optional, Any, Union
+import asyncio
+from dataclasses import dataclass
+import datetime
+import traceback
+
+import numpy as np
+import pandas as pd
+
+from gufe.protocols.protocolunit import (
+    ProtocolUnit,
+    ProtocolUnitResult,
+    ProtocolUnitFailure,
+    Context,
+)
+from gufe.settings import Settings
+
+from alchemiscale.models import ScopedKey
+from feflow.utils.data import deserialize
+
+from ..compute.client import FahAdaptiveSamplingClient
+from ..compute.models import JobStateEnum, FahProject, FahRun, FahClone
+from ..compute.index import FahComputeServiceIndex
+from ..utils import NONBONDED_EFFORT, NonbondedSettings
+
+
+class FahExecutionException(RuntimeError): ...
+
+
+@dataclass
+class FahContext(Context):
+    fah_client: FahAdaptiveSamplingClient
+    fah_projects: list[FahProject]
+    project_run_clone: Tuple[Optional[str], Optional[str], Optional[str]]
+    transformation_sk: ScopedKey
+    task_sk: ScopedKey
+    index: FahComputeServiceIndex
+    fah_poll_sleep: int = 60
+    encryption_public_key: Optional[str] = None
+
+
+class FahSimulationUnit(ProtocolUnit):
+
+    async def execute(
+        self, *, context: Context, raise_error: bool = False, **inputs
+    ) -> Union[ProtocolUnitResult, ProtocolUnitFailure]:
+        """Given `ProtocolUnitResult` s from dependencies, execute this `ProtocolUnit`.
+
+        Parameters
+        ----------
+        context : Context
+            Execution context for this `ProtocolUnit`; includes e.g. ``shared``
+            and ``scratch`` `Path` s.
+        raise_error : bool
+            If True, raise any errors instead of catching and returning a
+            `ProtocolUnitFailure` default False
+        **inputs
+            Keyword arguments giving the named inputs to `_execute`.
+            These can include `ProtocolUnitResult` objects from `ProtocolUnit`
+            objects this unit is dependent on.
+
+        """
+        result: Union[ProtocolUnitResult, ProtocolUnitFailure]
+        start = datetime.datetime.now()
+
+        try:
+            outputs = await self._execute(context, **inputs)
+            result = ProtocolUnitResult(
+                name=self.name,
+                source_key=self.key,
+                inputs=inputs,
+                outputs=outputs,
+                start_time=start,
+                end_time=datetime.datetime.now(),
+            )
+
+        except KeyboardInterrupt:
+            # if we "fail" due to a KeyboardInterrupt, we always want to raise
+            raise
```

Comment:
Raise something concrete here to distinguish from other error paths that have raw raise.

Eg. 

```python
        except KeyboardInterrupt as e
            # if we "fail" due to a KeyboardInterrupt, we always want to raise
            raise RuntimeError("Caught keyboard interrupt") from e
            
```